dashboard/app/asset_storage.go

```diff
@@ -0,0 +1,577 @@
+// Copyright 2022 syzkaller project authors. All rights reserved.
+// Use of this source code is governed by Apache 2 LICENSE that can be found in the LICENSE file.
+
+package main
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/google/syzkaller/dashboard/dashapi"
+	"github.com/google/syzkaller/pkg/assets"
+	"golang.org/x/net/context"
+	"google.golang.org/appengine/v2/datastore"
+	db "google.golang.org/appengine/v2/datastore"
+	"google.golang.org/appengine/v2/log"
+)
+
+/*
+   TODO: decide if we really want to save job-related assets right now.
+
+   The problem is that:
+   1) We cannot first report the job status and then upload the assets -- an email will already
+      be sent by that time.
+   2) If we first upload assets and then construct a JobDoneReq response already with assets,
+      we need to handle extra corner cases:
+        - If JobDoneReq failed, assets must be deleted.
+        - If we uploaded several files, all must be deleted.
+        - What if syz-ci was killed somewhere in between? How do we get rid of those abadoned files?
+
+   With the existing asset types it does not seem to be worth it to solve those problems.
+   Probably once we have things like code dumps it will start to make some sense.
+*/
+
+/*
+   TODO:
+   - Incorporate Assets into BugReport
+   - Include asset info into email reports. Important assets are already propagated.
+   - Integrate with the web dashboard
+*/
+
+/*
+   Currently the code below only supports build-specific assets.
+   At some point we'll also implement crash-specific assets. What needs to be done then:
+
+   1) Implement saveCrashAsset method. forgetAssets already works for crash-specific stuff.
+   2) Add support to the policies, currently only BugStatusDeprecationPolicy takes care of
+      crash-specific assets.
+   3) Adjust dashapi, tests and bug reporting.
+
+   TODO: how do we handle crash asset deprecation? Crashes are very easily obsoleted, and so
+   the contents of the Asset array will just be gone. Looks like we need to duplicate them
+   somewhere. Or use the alternative design, as described below.
+*/
+
+/*
+   Alternative DB design: don't embed assets, but store them as separate entities.
```

Comment:
Looking at how complex the deprecated asset detection gets when we embed Assets (see the implemented code), I'm leaning more and more towards this somewhat simpler approach. 

@dvyukov, what do you think?

Reply:
FTR after an offline discussion it was decided to switch from querying assets to forget to querying assets to remember. This leads to fewer corner cases.